libs/langchain/langchain/chat_models/fireworks.py

```diff
@@ -0,0 +1,140 @@
+    SystemMessageChunk,
+)
+from langchain.schema.output import ChatGeneration, ChatGenerationChunk, ChatResult
+import openai
+
+
+def _convert_delta_to_message_chunk(
+    _dict: Mapping[str, Any], default_class: type[BaseMessageChunk]
+) -> BaseMessageChunk:
+    role = _dict.get("role")
+    content = _dict.get("content") or ""
+    if _dict.get("function_call"):
+        additional_kwargs = {"function_call": dict(_dict["function_call"])}
+    else:
+        additional_kwargs = {}
+
+    if role == "user" or default_class == HumanMessageChunk:
+        return HumanMessageChunk(content=content)
+    elif role == "assistant" or default_class == AIMessageChunk:
+        return AIMessageChunk(content=content, additional_kwargs=additional_kwargs)
+    elif role == "system" or default_class == SystemMessageChunk:
+        return SystemMessageChunk(content=content)
+    elif role == "function" or default_class == FunctionMessageChunk:
+        return FunctionMessageChunk(content=content, name=_dict["name"])
+    elif role or default_class == ChatMessageChunk:
+        return ChatMessageChunk(content=content, role=role)
+    else:
+        return default_class(content=content)
+
+
+class ChatFireworks(BaseChatModel):
+    """Fireworks Chat models."""
+
+    model = "accounts/fireworks/models/llama-v2-13b-chat"
+    model_kwargs: Optional[dict] = {"temperature": 0.7, "max_tokens": 512, "top_p": 1}
+    fireworks_api_url: Optional[str] = "https://api.fireworks.ai/inference/v1"
+    fireworks_api_key: Optional[str] = os.environ.get("FIREWORKS_API_KEY")
+
+    @property
+    def _llm_type(self) -> str:
+        """Return type of llm."""
+        return "fireworks-chat"
+
+    def _generate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        message_dicts = self._create_message_dicts(messages, stop)
+        response = openai.ChatCompletion.create(
+            api_base=self.fireworks_api_url,
+            api_key=self.fireworks_api_key,
+            model=self.model,
+            messages=message_dicts,
+            **self.model_kwargs,
+        )
+        return self._create_chat_result(response)
+
+    async def _agenerate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[CallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        message_dicts = self._create_message_dicts(messages, stop)
+        response = await openai.ChatCompletion.acreate(
+            api_base=self.fireworks_api_url,
+            api_key=self.fireworks_api_key,
+            model=self.model,
+            messages=message_dicts,
+            **self.model_kwargs,
+        )
+        return self._create_chat_result(response)
+
+    def _combine_llm_outputs(self, llm_outputs: List[Optional[dict]]) -> dict:
```

Comment:
This implementation differs from that of ChatOpenAI. Is it by design?

Reply:
The difference would be I didn't record `token_usage`, and I am not quite sure if that is necessary.

Reply:
How does the output look like in our case? Anyscale is using the same output schema as OpenAI https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/chat_models/anyscale.py#L30

It looks like in the ChatOpenAI case, the output of this method contains only stats but we are including also the completions. I think the completions go to a separate field: https://github.com/langchain-ai/langchain/blob/master/libs/langchain/langchain/chat_models/base.py#L311

Could you share a sample output from ChatOpenAI.generate() and compare it against ChatFireworks.generate()? Lets make sure that the output schema is similar.

Reply:
ChatFireworks
```
generations=[[ChatGeneration(text="Hello! It's nice to meet you.", 
generation_info={'finish_reason': 'stop'}, 
message=AIMessage(content="Hello! It's nice to meet you.", additional_kwargs={}, example=False))]] 
llm_output={'model': 'accounts/fireworks/models/llama-v2-7b-chat'} 
run=[RunInfo(run_id=UUID('b73318cb-7951-41c1-a35c-81e439446cc0'))]
```

ChatOpenAI
```
generations=[[ChatGeneration(text='Hello! How can I assist you today?', 
generation_info={'finish_reason': 'stop'}, 
message=AIMessage(content='Hello! How can I assist you today?', additional_kwargs={}, example=False))]] llm_output={'token_usage': {'prompt_tokens': 8, 'completion_tokens': 9, 'total_tokens': 17}, 'model_name': 'gpt-3.5-turbo'} 
run=[RunInfo(run_id=UUID('8ec0dc0f-7129-4c2c-b37f-8bedd328aacb'))]
```

Reply:
lg. Thanks for validating. The current code is fine in this case.